base/promotion.jl

```diff
@@ -217,13 +217,27 @@ max(x::Real, y::Real) = max(promote(x,y)...)
 min(x::Real, y::Real) = min(promote(x,y)...)
 minmax(x::Real, y::Real) = minmax(promote(x, y)...)
 
-# "Promotion" that takes a function into account. You can override this
-# as needed. For example, if you need to provide a custom result type
-# for the multiplication of two types,
-#   promote_op{R<:MyType,S<:MyType}(::typeof(*), ::Type{R}, ::Type{S}) = MyType{multype(R,S)}
+# "Promotion" that takes a function into account. These are meant to be used
+# mainly by broadcast methods, so it is adviced against overriding them
 promote_op(::Any)    = (@_pure_meta; Bottom)
-promote_op(::Any, ::Any, ::Any...) = (@_pure_meta; Any)
 promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)
+promote_op{T}(::Type{T}, ::Type) = (@_pure_meta; T) # To handle ambiguities
+if isdefined(Core, :Inference)
+    function promote_op(op, T::Type)
+        G = Tuple{Generator{Tuple{T},typeof(op)}}
+        R = Core.Inference.return_type(first, G)
+        return isleaftype(R) ? R : Any
+    end
+    function promote_op(op, R::Type, S::Type)
+        isleaftype(R) || isleaftype(S) || return typejoin(R, S)
```

Comment:
There surely are. I guess this is where the "take the type of the actual elements" path would be needed. Then e.g. `Number[1] + Number[1]` would give `Int[2]` instead of `Number[2]` though, but that seems still better than `Any[2]`. (I guess this would be one of the critical situations?)